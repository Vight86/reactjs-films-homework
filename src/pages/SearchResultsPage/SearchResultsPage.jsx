import React from 'react';
import propTypes from 'prop-types';
import Header from '../../components/Header/index';
import MovieList from '../../components/MovieList/index';
import FilterList from '../../components/FilterList/index';
import Button from '../../components/Button/index';
import SelectButton from '../../components/SelectButton/index';
import ToggleButton from '../../components/ToggleButton/index';
import Preloader from '../../components/Preloader/index';
import styles from './SearchResultsPage.scss';

const SearchResultsPage = ({
  movies,
  genres,
  isGrid,
  handleGridToggle,
  movieTrailerKey,
  updateMovieTrailerKey,
  getTrending,
  getTopRatedMovies,
  getUpcoming,
  status,
  filterMovies,
}) => (
  <>
    <main className={styles.main}>
      <Header className="secondary">
        <FilterList>
          <Button className="filterButton" onClick={getTrending}>Trending</Button>
          <Button className="filterButton" onClick={getTopRatedMovies}>Top rated</Button>
          <Button className="filterButton" onClick={getUpcoming}>Coming soon</Button>
          <SelectButton genres={genres} filterMovies={filterMovies}>Genre</SelectButton>
        </FilterList>
        <ToggleButton className="toggleGridButton" isGrid={isGrid} handleGridToggle={handleGridToggle} />
      </Header>
      <MovieList
        movies={movies}
        genres={genres}
        isGrid={isGrid}
        movieTrailerKey={movieTrailerKey}
        updateMovieTrailerKey={updateMovieTrailerKey}
      />
    </main>
    { status === 'loading' ? <Preloader /> : null}
  </>
);

SearchResultsPage.propTypes = {
  movies: propTypes.arrayOf(propTypes.object).isRequired,
  genres: propTypes.arrayOf(propTypes.object).isRequired,
  isGrid: propTypes.bool.isRequired,
  handleGridToggle: propTypes.func.isRequired,
  movieTrailerKey: propTypes.string,
  updateMovieTrailerKey: propTypes.func.isRequired,
  getTrending: propTypes.func.isRequired,
  getTopRatedMovies: propTypes.func.isRequired,
  getUpcoming: propTypes.func.isRequired,
  filterMovies: propTypes.func.isRequired,
  status: propTypes.string.isRequired,
};

SearchResultsPage.defaultProps = {
  movieTrailerKey: '',
};

export default SearchResultsPage;
