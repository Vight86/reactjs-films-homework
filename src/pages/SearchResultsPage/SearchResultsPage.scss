@import '../../utils/variables';

.main {
  padding-right: $pr;
  padding-left: $pl;

  @media screen and (max-width: $bp-xxlarge) {
    padding-right: $pr-small;
    padding-left: $pl-small;
  }
}
