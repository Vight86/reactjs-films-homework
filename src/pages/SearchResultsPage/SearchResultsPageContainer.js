import React, { useState, useCallback, useEffect } from 'react';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';

import { apiPathUpdated, selectQuery } from '../../modules/query/index';
import {
  selectMovies, loadMovies, addMovies, currentMovieTrailerKeyAdded, addCurrentMovieTrailerKey,
} from '../../modules/movies/index';
import { loadGenres, selectGenres } from '../../modules/genres/index';
import { updateStatus, selectStatus } from '../../modules/status/index';
import { genreFilterUpdated, selectGenreId } from '../../modules/filters/index';

import SearchResultsPage from './SearchResultsPage';

const SearchResultsPageContainer = () => {
  const dispatch = useDispatch();

  const {
    basePath, apiPath, apiKey, lang, genresPath, adult, searchQuery,
  } = useSelector(selectQuery);

  const genres = useSelector(selectGenres, shallowEqual);
  useEffect(() => {
    const fetchGenresUrl = `${basePath}${genresPath}?api_key=${apiKey}&language=${lang}`;
    dispatch(loadGenres(fetchGenresUrl));
  }, []);

  const getFetchUrl = (query, pageNumber) => {
    let fetchMoviesUrl;
    if (query !== '') {
      fetchMoviesUrl = `${basePath}${apiPath}?api_key=${apiKey}&language=${lang}&query=${searchQuery}&page=${pageNumber}&include_adult=${adult}`;
    } else {
      fetchMoviesUrl = `${basePath}${apiPath}?api_key=${apiKey}&language=${lang}&page=${pageNumber}`;
    }

    return fetchMoviesUrl;
  };

  const { movies, pageSize } = useSelector(selectMovies, shallowEqual);
  let { page } = useSelector(selectMovies);
  const genreId = useSelector(selectGenreId);
  let moviesToRender = movies.filter((_, index) => index < page * pageSize);

  moviesToRender = (genreId === 0)
    ? moviesToRender
    : moviesToRender.filter((movie) => movie.genreIds.includes(genreId));

  useEffect(() => {
    page = 1;
    let fetchUrl = getFetchUrl(searchQuery, page);
    dispatch(loadMovies(fetchUrl));

    const handleScroll = () => {
      const windowRelativeBottom = document.documentElement.getBoundingClientRect().bottom;
      if (Math.trunc(windowRelativeBottom) <= document.documentElement.clientHeight + 5) {
        page += 1;
        fetchUrl = getFetchUrl(searchQuery, page);
        dispatch(addMovies(fetchUrl));
      }
    };
    window.addEventListener('scroll', handleScroll);

    if (page * pageSize >= movies.length && movies.length !== 0) {
      dispatch(updateStatus('loaded'));
    }
  }, [apiPath, searchQuery, genreId]);

  const [isGrid, setIsGrid] = useState(true);
  const handleGridToggle = () => {
    setIsGrid(!isGrid);
  };

  const { currentMovieTrailerKey } = useSelector(selectMovies, shallowEqual);
  const updateMovieTrailerKey = (id) => {
    if (id === null) {
      dispatch(currentMovieTrailerKeyAdded(id));
      return;
    }
    const url = `${basePath}/movie/${id}/videos?api_key=${apiKey}`;
    dispatch(addCurrentMovieTrailerKey(url));
  };

  const getTrending = useCallback(
    () => dispatch(apiPathUpdated('/movie/popular')),
    [dispatch],
  );

  const getTopRatedMovies = useCallback(
    () => dispatch(apiPathUpdated('/movie/top_rated')),
    [dispatch],
  );

  const getUpcoming = useCallback(
    () => dispatch(apiPathUpdated('/movie/upcoming')),
    [dispatch],
  );

  const filterMovies = useCallback(
    (id) => dispatch(genreFilterUpdated(id)),
    [dispatch],
  );

  const { status } = useSelector(selectStatus);
  return (
    <SearchResultsPage
      movies={moviesToRender}
      genres={genres}
      isGrid={isGrid}
      handleGridToggle={handleGridToggle}
      movieTrailerKey={currentMovieTrailerKey}
      updateMovieTrailerKey={updateMovieTrailerKey}
      getTrending={getTrending}
      getTopRatedMovies={getTopRatedMovies}
      getUpcoming={getUpcoming}
      filterMovies={filterMovies}
      status={status}
    />
  );
};

export default SearchResultsPageContainer;
