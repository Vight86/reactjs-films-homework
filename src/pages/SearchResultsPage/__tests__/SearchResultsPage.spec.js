import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import SearchResultsPage from '../index';

const setUp = () => {
  const renderer = new ShallowRenderer();
  renderer.render(<SearchResultsPage />);
  return renderer.getRenderOutput();
};

describe('render SearchResultsPage component', () => {
  it('should be main tag', () => {
    const result = setUp();
    expect(result.type).toBe('main');
  });

  it('render component without props', () => {
    const result = setUp();
    expect(result).toMatchSnapshot();
  });
});
