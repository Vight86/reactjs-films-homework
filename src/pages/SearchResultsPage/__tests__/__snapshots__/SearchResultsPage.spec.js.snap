// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render SearchResultsPage component render SearchResultsPage grid view correctly 1`] = `
<React.Fragment>
  <main
    className="main"
  >
    <Header
      className="secondary"
    >
      <FilterList>
        <Button
          className="filterButton"
          onClick={[MockFunction]}
        >
          Trending
        </Button>
        <Button
          className="filterButton"
          onClick={[MockFunction]}
        >
          Top rated
        </Button>
        <Button
          className="filterButton"
          onClick={[MockFunction]}
        >
          Coming soon
        </Button>
        <SelectButton
          filterMovies={[MockFunction]}
          genres={
            Array [
              Object {
                "id": 1,
                "name": "Action",
              },
              Object {
                "id": 2,
                "name": "SciFi",
              },
            ]
          }
        >
          Genre
        </SelectButton>
      </FilterList>
      <ToggleButton
        className="toggleGridButton"
        handleGridToggle={[MockFunction]}
        isGrid={true}
      />
    </Header>
    <MovieList
      genres={
        Array [
          Object {
            "id": 1,
            "name": "Action",
          },
          Object {
            "id": 2,
            "name": "SciFi",
          },
        ]
      }
      isGrid={true}
      movieTrailerKey="testMovieTrailerKey"
      movies={
        Array [
          Object {
            "adult": false,
            "backdropPath": "/test.jpg",
            "genreIds": Array [
              1,
              2,
            ],
            "id": 1,
            "originalLanguage": "en",
            "originalTitle": "test",
            "overview": "test",
            "popularity": 1,
            "posterPath": "/test.jpg",
            "releaseDate": "1",
            "title": "test",
            "video": false,
            "voteAverage": 1,
            "voteCount": 1,
          },
          Object {
            "adult": false,
            "backdropPath": "/test.jpg",
            "genreIds": Array [
              1,
              2,
            ],
            "id": 1,
            "originalLanguage": "en",
            "originalTitle": "test",
            "overview": "test",
            "popularity": 1,
            "posterPath": "/test.jpg",
            "releaseDate": "1",
            "title": "test",
            "video": false,
            "voteAverage": 1,
            "voteCount": 1,
          },
        ]
      }
      updateMovieTrailerKey={[MockFunction]}
    />
  </main>
</React.Fragment>
`;

exports[`render SearchResultsPage component render SearchResultsPage table view correctly 1`] = `
<React.Fragment>
  <main
    className="main"
  >
    <Header
      className="secondary"
    >
      <FilterList>
        <Button
          className="filterButton"
          onClick={[MockFunction]}
        >
          Trending
        </Button>
        <Button
          className="filterButton"
          onClick={[MockFunction]}
        >
          Top rated
        </Button>
        <Button
          className="filterButton"
          onClick={[MockFunction]}
        >
          Coming soon
        </Button>
        <SelectButton
          filterMovies={[MockFunction]}
          genres={
            Array [
              Object {
                "id": 1,
                "name": "Action",
              },
              Object {
                "id": 2,
                "name": "SciFi",
              },
            ]
          }
        >
          Genre
        </SelectButton>
      </FilterList>
      <ToggleButton
        className="toggleGridButton"
        handleGridToggle={[MockFunction]}
        isGrid={false}
      />
    </Header>
    <MovieList
      genres={
        Array [
          Object {
            "id": 1,
            "name": "Action",
          },
          Object {
            "id": 2,
            "name": "SciFi",
          },
        ]
      }
      isGrid={false}
      movieTrailerKey="testMovieTrailerKey"
      movies={
        Array [
          Object {
            "adult": false,
            "backdropPath": "/test.jpg",
            "genreIds": Array [
              1,
              2,
            ],
            "id": 1,
            "originalLanguage": "en",
            "originalTitle": "test",
            "overview": "test",
            "popularity": 1,
            "posterPath": "/test.jpg",
            "releaseDate": "1",
            "title": "test",
            "video": false,
            "voteAverage": 1,
            "voteCount": 1,
          },
          Object {
            "adult": false,
            "backdropPath": "/test.jpg",
            "genreIds": Array [
              1,
              2,
            ],
            "id": 1,
            "originalLanguage": "en",
            "originalTitle": "test",
            "overview": "test",
            "popularity": 1,
            "posterPath": "/test.jpg",
            "releaseDate": "1",
            "title": "test",
            "video": false,
            "voteAverage": 1,
            "voteCount": 1,
          },
        ]
      }
      updateMovieTrailerKey={[MockFunction]}
    />
  </main>
</React.Fragment>
`;

exports[`render SearchResultsPage component render loading status correctly 1`] = `
<React.Fragment>
  <main
    className="main"
  >
    <Header
      className="secondary"
    >
      <FilterList>
        <Button
          className="filterButton"
          onClick={[MockFunction]}
        >
          Trending
        </Button>
        <Button
          className="filterButton"
          onClick={[MockFunction]}
        >
          Top rated
        </Button>
        <Button
          className="filterButton"
          onClick={[MockFunction]}
        >
          Coming soon
        </Button>
        <SelectButton
          filterMovies={[MockFunction]}
          genres={
            Array [
              Object {
                "id": 1,
                "name": "Action",
              },
              Object {
                "id": 2,
                "name": "SciFi",
              },
            ]
          }
        >
          Genre
        </SelectButton>
      </FilterList>
      <ToggleButton
        className="toggleGridButton"
        handleGridToggle={[MockFunction]}
        isGrid={false}
      />
    </Header>
    <MovieList
      genres={
        Array [
          Object {
            "id": 1,
            "name": "Action",
          },
          Object {
            "id": 2,
            "name": "SciFi",
          },
        ]
      }
      isGrid={false}
      movieTrailerKey="testMovieTrailerKey"
      movies={
        Array [
          Object {
            "adult": false,
            "backdropPath": "/test.jpg",
            "genreIds": Array [
              1,
              2,
            ],
            "id": 1,
            "originalLanguage": "en",
            "originalTitle": "test",
            "overview": "test",
            "popularity": 1,
            "posterPath": "/test.jpg",
            "releaseDate": "1",
            "title": "test",
            "video": false,
            "voteAverage": 1,
            "voteCount": 1,
          },
          Object {
            "adult": false,
            "backdropPath": "/test.jpg",
            "genreIds": Array [
              1,
              2,
            ],
            "id": 1,
            "originalLanguage": "en",
            "originalTitle": "test",
            "overview": "test",
            "popularity": 1,
            "posterPath": "/test.jpg",
            "releaseDate": "1",
            "title": "test",
            "video": false,
            "voteAverage": 1,
            "voteCount": 1,
          },
        ]
      }
      updateMovieTrailerKey={[MockFunction]}
    />
  </main>
  <Preloader />
</React.Fragment>
`;
